### PR TITLE
Improve quest drop notification

### PR DESCRIPTION
**Pull request by SabreCat**
1. Changes the notification of quests awarded on levelup to use a modal instead of a drop-style alert.
2. Fixes (I think!) the bug where quest scroll drops appeared twice, then reset to 1 in inventory upon refresh.
3. Refactors the quests Jade file to use a mixin for reused quest information readout.
4. ~~Refactors drop notifications to use `switch` instead of a bunch of `else if`s.~~

### CODE REVIEW

File: website/views/shared/modals/quests.jade [hunk 1]
```diff
@@ -90,3 +77,17 @@ script(type='text/ng-template', id='modals/questInvitation.html')
...
+    h4=env.t('messageDropQuest')
+  .modal-body
+    h4 {{::selectedQuest.text()}}
+    +questInfo
+    hr
+    .npc_ian.pull-left
+    p=env.t('dropQuestCongrats')
+  .modal-footer
+    button.btn.btn-default(ng-click='closeQuest(); $close()')=env.t('questLater')
+    button.btn.btn-primary(ng-click='questInit(); $close()')=env.t('inviteParty')
```
Comment (SabreCat) [inline website/views/shared/modals/quests.jade]: I wanted to have this conditionally be the 'create a party' button, but `inviteOrStartParty` wasn't working here.

File: website/public/js/controllers/notificationCtrl.js [hunk 2]
```diff
@@ -58,9 +58,11 @@ habitrpg.controller('NotificationCtrl',
 
     $rootScope.$watch('user._tmp.drop', function(after, before){
       // won't work when getting the same item twice?
-      if (after == before || !after) return;
-      $rootScope.playSound('Achievement_Unlocked');
-      if (after.type !== 'gear') {
+      if (_.isEqual(after, before) || !after) return;
```
Comment (gisikw) [inline website/public/js/controllers/notificationCtrl.js:61]: Why not `===`?

File: website/public/js/controllers/notificationCtrl.js [hunk 3]
```diff
@@ -70,23 +72,30 @@ habitrpg.controller('NotificationCtrl',
...
-        var notes = Content.eggs[after.key].notes();
-        Notification.drop(env.t('messageDropEgg', {dropText: text, dropNotes: notes}), after);
-      }else if(after.type === 'Food'){
-        var text = Content.food[after.key].text();
-        var notes = Content.food[after.key].notes();
-        Notification.drop(env.t('messageDropFood', {dropArticle: after.article, dropText: text, dropNotes: notes}), after);
-      }else{
-        // Keep support for another type of drops that might be added
-        Notification.drop(User.user._tmp.drop.dialog);
+      switch(after.type) {
```
Comment (gisikw) [inline website/public/js/controllers/notificationCtrl.js]: Booo switch statements (honestly, I'd prefer we avoid switch statements. It's far too easy to forget a break and failthrough. Especially with OSS).

Comment (SabreCat) [inline website/public/js/controllers/notificationCtrl.js]: Ah, if that's a bad pattern, I can revert it. I find it more readable than an `else if` chain, but I'm just a scrub, heh.

File: website/public/js/controllers/notificationCtrl.js (hunk 2 above)
Comment (SabreCat) [inline website/public/js/controllers/notificationCtrl.js:61]: It wasn't working, in mysterious fashion. I would see the `if` statement blow right by, despite both `after` and `before` being, e.g. {type:'Quest',key:'atom1'}. Changing it to `_.isEqual` stopped that happening, eliminating the double-notification bug.

File: website/public/js/controllers/notificationCtrl.js [hunk 4]
```diff
@@ -58,9 +58,11 @@ habitrpg.controller('NotificationCtrl',
...
       // won't work when getting the same item twice?
-      if (after == before || !after) return;
-      $rootScope.playSound('Achievement_Unlocked');
-      if (after.type !== 'gear') {
+      if (_.isEqual(after, before) || !after) return;
+      var text, notes;
+      $rootScope.playSound('Item_Drop');
+
+      if (after.type !== 'gear' && after.type !== 'Quest') {
         var type = (after.type == 'Food') ? 'food' :
```
Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js]: IMO, once ternaries turn into double ternaries, it's time to go with something that is not a ternary.

File: website/views/shared/modals/quests.jade [hunk 5]
```diff
@@ -1,5 +1,23 @@
 include ./quest-rewards
 
+mixin questInfo
```
Comment (crookedneighbor) [inline website/views/shared/modals/quests.jade:3]: so pleased about this. :heart:

File: website/public/js/controllers/notificationCtrl.js (hunk 3 above)
Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js]: I agree that it's easier to parse, but ultimately a bad pattern.

Comment (gisikw) [inline website/public/js/controllers/notificationCtrl.js]: One alternative is the hashdict lookup:

``` js
handling = {
  Quest: function() {
    $rootScope.selectedQuest = Content.quests[after.key];
    $rootScope.openModal('questDrop');
  }
  // etc
};

handling[after.type]();
```

But that would probably require a bit of refactoring.

File: website/public/js/controllers/notificationCtrl.js (hunk 4 above)
Comment (SabreCat) [inline website/public/js/controllers/notificationCtrl.js]: OK, I'll refactor that into another `if`/`else if` chain.

File: website/public/js/controllers/notificationCtrl.js [hunk 6]
```diff
@@ -58,35 +58,42 @@ habitrpg.controller('NotificationCtrl',
...
+      if (_.isEqual(after, before) || !after) return;
+      var text, notes, type;
+      $rootScope.playSound('Item_Drop');
+
+      if (after.type !== 'gear' && after.type !== 'Quest') {
+        if (after.type === 'Food') {
+          type = 'food';
+        } else if (after.type === 'HatchingPotion') {
+          type = 'hatchingPotions';
+        } else type = after.type.toLowerCase() + 's';
```
Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js]: Doesn't food fall into this?

Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js]: Oh never mind. Foods

Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js]: Id prefer that the last else had brackets.

File: website/public/js/controllers/notificationCtrl.js [hunk 7]
```diff
@@ -58,35 +58,42 @@ habitrpg.controller('NotificationCtrl',
...
         if(!User.user.items[type][after.key]){
           User.user.items[type][after.key] = 0;
         }
         User.user.items[type][after.key]++;
       }
 
-      if(after.type === 'HatchingPotion'){
```
Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js:73]: I think we can do some meta programming here to mitigate this awful if/else block, but that can be a refactor we do after this is merged in.

File: website/public/js/controllers/notificationCtrl.js [hunk 8]
```diff
@@ -121,17 +128,6 @@ habitrpg.controller('NotificationCtrl',
       $rootScope.openModal('achievements/contributor');
     });
 
-    /*_.each(['weapon', 'head', 'chest', 'shield'], function(watched){
```
Comment (crookedneighbor) [inline website/public/js/controllers/notificationCtrl.js:124]: Death to commented out code!

File: website/views/shared/modals/quests.jade (hunk 1 above)
Comment (crookedneighbor) [inline website/views/shared/modals/quests.jade]: I don't think you're passing in a controller when opening this modal. Would it be helpful to pass in the partyCtrl and create a function on `$scope` there that does the `inviteOrStartParty` code? This is how it's done in the Header controller: https://github.com/HabitRPG/habitrpg/blob/32a74471dbeb791fe5fafc6d5441e5cda827968b/website/public/js/controllers/headerCtrl.js#L9

Comment (SabreCat) [inline website/views/shared/modals/quests.jade]: I can give that a try!

File: website/public/js/controllers/notificationCtrl.js (hunk 8 above)
Comment (SabreCat) [inline website/public/js/controllers/notificationCtrl.js:124]: You're like a one-man angry mob, calling for the death of various things that displease you!

Comment (gisikw) [inline website/public/js/controllers/notificationCtrl.js:124]: Death to commented out code!

Comment (gisikw) [inline website/public/js/controllers/notificationCtrl.js:124]: (commented code is a artifact from pre-version-control days)

File: website/public/js/services/questServices.js [hunk 9]
```diff
@@ -84,12 +85,22 @@
...
+    }
+
     return {
       lockQuest: lockQuest,
       buyQuest: buyQuest,
       questPopover: questPopover,
       showQuest: showQuest,
-      closeQuest: closeQuest
+      closeQuest: closeQuest,
+      questInit: questInit
```
Comment (SabreCat) [inline website/public/js/services/questServices.js:103]: Implementing the Invite Party button on the quest drop modal involved moving another function out of the controllers into the Quests service. How does this look?

File: website/public/js/services/questServices.js [hunk 10]
```diff
@@ -84,12 +85,22 @@
       $rootScope.selectedQuest = undefined;
     }
 
+    function questInit(){
+      Analytics.track({'hitType':'event','eventCategory':'behavior','eventAction':'quest','owner':true,'response':'accept','questName':$rootScope.selectedQuest.key});
+      $rootScope.party.$questAccept({key:$rootScope.selectedQuest.key}, function(){
+        $rootScope.party.$get();
+      });
+      $rootScope.$state.go('options.social.party');
```
Comment (SabreCat) [inline website/public/js/services/questServices.js]: Inviting the party to your quest had a curious lack of user feedback; I added this so we jump to the Party page. Does that look right?

**Pull request merged**